src/metaswitch/clearwater/config_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,144 @@
+            value = self.read_from_etcd()
+            if self._terminate_flag:
+                break
+
+            if len(value) > 0:
+                _log.debug("Got new config value from etcd:\n{}".format(value))
+                self._plugin.on_config_changed(value)
+                self._alarm.update_file(self._plugin.file())
+
+    # Read the current value of the key from etcd (blocks until there's a
+    # change).
+    def read_from_etcd(self):
+        value = None
+        try:
+            result = self._client.get(self._key)
+
+            # If the key hasn't changed since we last saw it, then
+            # wait for it to change before doing anything else.
+            _log.info("Read config value for {} from etcd (epoch {})".format(
+                        self._key,
+                        result.modifiedIndex))
+
+            if result.modifiedIndex == self._index:
+                while not self._terminate_flag:
+                    try:
+                        result = self._client.watch(self._key,
+                                                    index=result.modifiedIndex+1,
+                                                    timeout=5,
+                                                    recursive=False)
+                        break
+                    except urllib3.exceptions.TimeoutError:
+                        # Timeouts after 5 seconds are expected, so ignore them
+                        # - unless we're terminating, we'll stay in the while
+                        # loop and try again
+                        pass
+                    except etcd.EtcdException as e:
+                        # We have seen timeouts getting raised as EtcdExceptions
+                        # so catch these here too and treat them as timeouts if
+                        # they indicate that the read timed out.
+                        if "Read timed out" in e.message:
+                            pass
+                        else:
+                            raise
+                    except ValueError:
+                        # The index isn't valid to watch on, probably because
+                        # there has been a snapshot between the get and the
+                        # watch. Just start the read again.
+                        _log.info("etcd index {} is invalid, retrying".format(
+                            result.modifiedIndex+1))
+                        self.read_from_etcd()
+
+                # Return if we're terminating.
+                if self._terminate_flag:
+                    return
+
+            # Save off the index of the result we're using for when we write
+            # back to etcd later.
+            self._index = result.modifiedIndex
+            value = result.value
+
+        except etcd.EtcdKeyError:
+            # If the key doesn't exist in etcd then there is currently no
+            # config.
+            self._index = None
+            pass
+        except Exception as e:
```

Comment:
I'm a bit worried by how general this exception handler is. It doesn't just catch exception caused by external interactions, it also catches code bugs (for which better behavior might be to propagate the exception or exit the process).